### PR TITLE
adjust blueprint generator scheduling

### LINKED ISSUE #25444: angular & language generators combined in a blueprint throws error during unit testing
##### **Overview of the issue**
When a blueprint include both angular & language generators, the unit testing throws the following error,

 ```
FAIL  generators/angular/generator.spec.js > SubGenerator angular of lnfix JHipster blueprint > run
TypeError: getWebappTranslation is not a function
 ❯ getTranslationValue ../../generator-jhipster/dist/generators/languages/support/translate.js:21:12

 ❯ replaceTranslationKeysWithText ../../generator-jhipster/dist/generators/languages/support/translate.js:56:29
 ❯ ../../generator-jhipster/dist/generators/languages/support/translate.js:75:106
 ❯ replaceAngularTranslations ../../generator-jhipster/dist/generators/angular/support/translate-angular.js:108:23
 ❯ ../../generator-jhipster/dist/generators/angular/support/translate-angular.js:127:37
 ❯ Transform.<anonymous> ../../generator-jhipster/node_modules/@yeoman/transform/src/transform.ts:78:13
 ❯ Transform.transform [as _transform] ../../generator-jhipster/node_modules/@yeoman/transform/src/transform.ts:41:29
```

Bu the same works fine while generating application.

##### **Motivation for or Use Case**
It should be possible to create a blueprint combining angular & language generators.

##### **Reproduce the error**
Use the attached .yo-rc.json

##### **Related issues**
None

##### **Suggest a Fix**
This issue is caused because, the blueprint:angular gets higher priority than the blueprint:language. Based on the hierarchy language is required before the frontend generators. Change the compose with in the language generator for blueprint to set the schedule property to false.

##### **JHipster Version(s)**
8.1.0

##### **JHipster configuration**
<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "additionalSubGenerators": "",
    "baseName": "lnfix",
    "cli": true,
    "entities": [],
    "generators": {
      "angular": {
        "command": true,
        "priorities": [
          "initializing",
          "prompting",
          "configuring",
          "composing",
          "loading",
          "preparing",
          "configuringEachEntity",
          "loadingEntities",
          "preparingEachEntity",
          "preparingEachEntityField",
          "preparingEachEntityRelationship",
          "postPreparingEachEntity",
          "default",
          "writing",
          "writingEntities",
          "postWriting",
          "postWritingEntities",
          "loadingTranslations",
          "install",
          "postInstall",
          "end"
        ],
        "sbs": false,
        "written": true
      },
      "languages": {
        "command": true,
        "priorities": [
          "initializing",
          "prompting",
          "configuring",
          "composing",
          "loading",
          "preparing",
          "configuringEachEntity",
          "loadingEntities",
          "preparingEachEntity",
          "preparingEachEntityField",
          "preparingEachEntityRelationship",
          "postPreparingEachEntity",
          "default",
          "writing",
          "writingEntities",
          "postWriting",
          "postWritingEntities",
          "loadingTranslations",
          "install",
          "postInstall",
          "end"
        ],
        "sbs": false,
        "written": true
      }
    },
    "jhipsterVersion": "8.1.0",
    "localBlueprint": false,
    "sampleWritten": true,
    "subGenerators": [
      "angular",
      "languages"
    ]
  }
}
</pre>
</details>

##### **Environment and Tools**

openjdk version "17.0.8" 2023-07-18
OpenJDK Runtime Environment Temurin-17.0.8+7 (build 17.0.8+7)
OpenJDK 64-Bit Server VM Temurin-17.0.8+7 (build 17.0.8+7, mixed mode, sharing)

git version 2.34.1

node: v18.19.0
npm: 10.2.3

Docker version 25.0.3, build 4debf41

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Browsers and Operating System**

- [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by dwarakaprasad**
fixes #25444

A full blueprint (sbsBlueprint - false) will schedule immediately i.e. replacing the queue **slot** of the blueprinted generator.
A sbs blueprint (sbsBlueprint - true) will **not** affect the blueprinted generator's slot in queue but will take its own slot down below it.

---

Please make sure the below checklist is followed for Pull Requests.

- [X] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [X] Tests are added where necessary
- [X] The JDL part is updated if necessary
- [X] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [X] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
This indeed is the correct behavior for full blueprints, but will break sbs blueprints.
Blueprint tasks will be queued before the blueprinted generator. sbs needs to be queued after.

The `schedule` parameter needs to be conditional, so it must be a function.
Should change here:
https://github.com/yeoman/environment/blob/1f2dd7bdcb9f9e58ed473c03a38afa0a58e63ecb/src/environment-base.ts#L428
to `schedule: typeof schedule === ' function' ? schedule(generatorInstance) : schedule`.

And add the conditional function before:
https://github.com/jhipster/generator-jhipster/blob/59a178a8cd6cd6cfddad93f1f92610bb792bba3f/generators/base/generator.ts#L586

**Comment by dwarakaprasad** [comment]
> This indeed is the correct behavior for full blueprints, but will break sbs blueprints. Blueprint tasks will be queued before the blueprinted generator. sbs needs to be queued after.
> 
> The `schedule` parameter needs to be conditional, so it must be a function. Should change here: https://github.com/yeoman/environment/blob/1f2dd7bdcb9f9e58ed473c03a38afa0a58e63ecb/src/environment-base.ts#L428 to `schedule: typeof schedule === ' function' ? schedule(generatorInstance) : schedule`.
> 
> And add the conditional function before:
> 
> https://github.com/jhipster/generator-jhipster/blob/59a178a8cd6cd6cfddad93f1f92610bb792bba3f/generators/base/generator.ts#L586

Thank you for your insight!

**please correct me if i am wrong.**
So to be on the same page,
```
blueprint = jhipster-<blueprint>:** namespace
blueprinted generator = jhipster:** namespace
A full blueprint is one where the sbsBlueprint is set to false
A full blueprint is one where the sbsBlueprint is set to true
```

As is today,
1. A full blueprint's task should be queued first followed by the blueprinted generator's task i.e. should be scheduled immediately instead of going tro the **environment:run** loop. 
`Currently composewithBlueprint method doesn't set the schedule value and so is defaulted to true and so all blueprints (full/sbs) are always being scheduled as of today. (not as expected)`

3. A sbs blueprint on the other hand should let its blueprinted generator's task to queue first and then get queued i.e. using the  _environment:run_ loop.
`same as above, sbs is always being scheduled as of today (same as expected)`

Expectation is,
1. A full blueprint should not be scheduled rather be queued immediately - so this mean schedule = false
2. A sbs blueprint should be scheduled - so schedule = true

Instead of only having a binary possibility, you are proposing to have a function that will let generators make sophisticated decisions. 

Since this proposal needs change to yeoman-environment, both releases will have to be timed.

**Comment by mshima** [comment]
Yes, that's it.

The only part that is not correct is:

> 1. queued first followed by the blueprinted generator's task

The blueprinted generator's tasks are not queue unless the full blueprint inherits its tasks. The composed blueprint is `halt`
Since the full blueprint completely overrides the blueprinted generator, it should replace its queue `slot` by queueing immediately too.

> Instead of only having a binary possibility, you are proposing to have a function that will let generators make sophisticated decisions.

Should be:
`schedule: generator => generator.sbsBlueprint`

> Since this proposal needs change to yeoman-environment, both releases will have to be timed.

No problem with that.
Are you willing to create the PR?

**Comment by dwarakaprasad** [comment]
> it should replace its queue `slot` by queueing immediately too.

Yes, it makes sense.


> Are you willing to create the PR?

Yes...

**Comment by dwarakaprasad** [comment]
@mshima changes made

1. @yeoman/type [changes](https://github.com/yeoman/yeoman-api/pull/6)
2. environment [change](https://github.com/yeoman/environment/pull/521)
3. jhipster [change](https://github.com/jhipster/generator-jhipster/pull/25445/commits/69276a8b905f2e47637521a48263c2e2f714c4c8)

File: generators/languages/generator.js [hunk 1]
```diff
@@ -70,6 +70,8 @@ export default class LanguagesGenerator extends BaseApplicationGenerator {
       this.supportedLanguages = supportedLanguages;
       this.composedBlueprints = await this.composeWithBlueprints('languages', {
         generatorArgs: this.options.languages,
+        // This is needed to prioritize languages generator over client and server generators.
+        schedule: false,
```
Comment (mshima) [inline generators/languages/generator.js:74]:
```suggestion
```

**Review by mshima — APPROVED** [review]

**Pull request merged**